fix(dev-infra): ignore comments when validating commit messages

gkalpak opened the pull request:
When creating a commit with the git cli, git pre-populates the editor used to enter the commit message with some comments (i.e. lines starting with `#`). These comments contain helpful instructions or information regarding the changes that are part of the commit. As happens with all commit message comments, they are removed by git and do not end up in the final commit message.

However, the file that is passed to the `commit-msg` to be validated still contains these comments. This may affect the outcome of the commit message validation. In such cases, the author will not realize that the commit message is not in the desired format until the linting checks fail on CI (which validates the final commit messages and is not affected by this issue), usually several minutes later.

Possible ways in which the commit message validation outcome can be affected:
- The minimum body length check may pass incorrectly, even if there is no actual body, because the comments are counted as part of the body.
- The maximum line length check may fail incorrectly due to a very long line in the comments.

This commit fixes the problem by removing comment lines before validating a commit message that comes from a file.

Fixes #37865.

ichi0609 commented:
h

ichi0609 commented:
[](https://t.me/y01_jp)

josephperrott commented on dev-infra/commit-message/validate-file.ts:
```diff
@@ -15,7 +15,13 @@ import {validateCommitMessage} from './validate';
+  // Read the commit message from the specified file and remove any comments (i.e. lines starting
+  // with `#`). Comments are automatically removed by git and are not part of the final commit
+  // message.
+  const commitMessage = readFileSync(resolve(getRepoBaseDir(), filePath), 'utf8')
```
Rather than removing the lines during the file reading step, should we update the commit message parser to drop the comment lines?

gkalpak replied:
Since removing these comment lines should only be relevant when reading this commit from a file (i.e. in the `validateFile()` function), I wanted to keep the changes as focused as possible. Both ways should work, though. Happy to go either way.

josephperrott replied:
I think its worth just doing in the parser since we should always ignore the comments when checking validation regardless of their origin.

gkalpak replied:
Updated. PTAL.

gkalpak committed:
fix(dev-infra): ignore comments when validating commit messages

When creating a commit with the git cli, git pre-populates the editor
used to enter the commit message with some comments (i.e. lines starting
with `#`). These comments contain helpful instructions or information
regarding the changes that are part of the commit. As happens with all
commit message comments, they are removed by git and do not end up in
the final commit message.

However, the file that is passed to the `commit-msg` to be validated
still contains these comments. This may affect the outcome of the commit
message validation. In such cases, the author will not realize that the
commit message is not in the desired format until the linting checks
fail on CI (which validates the final commit messages and is not
affected by this issue), usually several minutes later.

Possible ways in which the commit message validation outcome can be
affected:
- The minimum body length check may pass incorrectly, even if there is
  no actual body, because the comments are counted as part of the body.
- The maximum line length check may fail incorrectly due to a very long
  line in the comments.

This commit fixes the problem by removing comment lines before
validating a commit message.

Fixes #37865

gkalpak committed:
fixup! fix(dev-infra): ignore comments when validating commit messages

josephperrott reviewed (approved):
LGTM
